[nrfconnect] Added support for DFU in light and lock examples. (#6575)
project-chip/connectedhomeip: examples/lighting-app/nrfconnect/README.md

device. As a result, device is able to join the Thread network and communicate
with other Thread devices in the network.

### Device Firmware Upgrade

> The example allows enabling Over-the-air Device Firmware Upgrade feature. In
> this process device hosting new firmware image sends it to the CHIP device using
> Bluetooth LE transport and
> [Simple Management Protocol](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/device_mgmt/index.html#device-mgmt).
> Then the old firmware image is replaced with the new one using
> [MCUboot](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/mcuboot/index.html)
> bootloader solution.

Reviewer: ```suggestion
The example allows enabling the over-the-air Device Firmware Upgrade feature. In
this process, the device hosting new firmware image sends the image to the CHIP device using
Bluetooth LE transport and
[Simple Management Protocol](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/device_mgmt/index.html#device-mgmt).
The [MCUboot](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/mcuboot/index.html) bootloader solution then replaces the old firmware image with the new one.
```

#### Bootloader

MCUboot is a secure bootloader used for swapping firmware images of different
versions and generating proper build output files that can be used in the device
firmware upgrade process.

Bootloader solution requires certain area of flash memory to perform application
images swapping process regarding the firmware upgrade. The Nordic devices use
for that purpose external memory chip that communicates with the microcontroller
via QSPI bus.

See the
[Building with Device Firmware Upgrade support](#building-with-device-firmware-upgrade-support)
section to learn how to change MCUboot and flash configuration in this example.

#### Simple Management Protocol

Simple Management Protocol (SMP) is a basic transfer encoding that is used for
device management purposes, including application image management. SMP supports

> using different transports like Bluetooth LE, UDP or serial USB/UART.

Reviewer: ```suggestion
using different transports, such as Bluetooth LE, UDP, or serial USB/UART.
```

> In this example CHIP device runs SMP Server in order to download application
> update image using Bluetooth LE transport.

Reviewer: ```suggestion
In this example, the CHIP device runs the SMP Server to download the application
update image using the Bluetooth LE transport.
```

See the
[Building with Device Firmware Upgrade support](#building-with-device-firmware-upgrade-support)

> section to learn how to enable SMP and utilize it for Device Firmware Upgrade
> purposes in this example.

Reviewer: ```suggestion
section to learn how to enable SMP and use it for the DFU purpose in this example.
```

    $ west build -b build-target -- -DOVERLAY_CONFIG=rpc.overlay

### Building with Device Firmware Upgrade support

> To build the example with configuration enabling DFU, run the following command

Reviewer: ```suggestion
To build the example with configuration that enables DFU, run the following command
```

with _build-target_ replaced with the build target name of the Nordic

> Semiconductor's kit you own (e.g.`nrf52840dk_nrf52840`):

Reviewer: e.g. > for example,

> **_WARNING:_** Please do remember about replacing _build-target_ also in the
> PM_STATIC_YML_FILE path.

    $ west build -b build-target -- -DOVERLAY_CONFIG=third_party/connectedhomeip/config/nrfconnect/app/overlay-dfu_support.conf -DPM_STATIC_YML_FILE="configuration/build-target/pm_static.yml"

#### Changing bootloader configuration

> To change default MCUboot configuration you can edit overlay file containing
> bootloader configuration options and located in the
> `config/nrfconnect/app/overlay-dfu_support.conf` or define desired options in
> your example `prj.conf` file.

Reviewer: ```suggestion
To change the default MCUboot configuration, edit the `overlay-dfu_support.conf` overlay file that contains
bootloader configuration options. The file is located in the
`config/nrfconnect/app` directory. You can also define the desired options in
your example's `prj.conf` file.
```

> It is important to introduce the same configuration changes, that were described
> above, in the `child_image/mcuboot.conf` file. That is necessary, as the
> bootloader image is separate application from the User application and it has
> own configuration file, which content must be consistent with the application
> configuration.

Reviewer: ```suggestion
Make sure to apply the same configuration changes in the `child_image/mcuboot.conf` file.
This is necessary for the configuration to work, as the bootloader image is a separate application from the user application and it has its own configuration file. The contents of this file must be consistent with the application
configuration.
```

#### Changing flash memory settings

> In the default configuration MCUboot uses
> [Partition Manager](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/scripts/partition_manager/partition_manager.html#partition-manager)
> to configure flash partitions utilized for the bootloader application image
> slots purposes. Those settings can be changed by defining
> [static partitions](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/scripts/partition_manager/partition_manager.html#ug-pm-static),
> what was done in this example case in order to define using external flash.

Reviewer: ```suggestion
In the default configuration, the MCUboot uses
the [Partition Manager](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/scripts/partition_manager/partition_manager.html#partition-manager)
to configure flash partitions used for the bootloader application image
slot purposes. You can change these settings by defining
[static partitions](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/scripts/partition_manager/partition_manager.html#ug-pm-static).
This example uses this option to define using an external flash.
```

> To modify flash settings of your board named by the _build-target_ (e.g.
> `nrf52840dk_nrf52840`) edit file located in
> `configuration/build-target/pm_static.yml`.

Reviewer: ```suggestion
To modify the flash settings of your board (that is, your _build-target_, for example
`nrf52840dk_nrf52840`), edit the `pm_static.yml` file located in
the `configuration/build-target/` directory.
```

to see how to use [CHIPTool](../../../src/android/CHIPTool/README.md) for
Android smartphones to commission and control the application within a
CHIP-enabled Thread network.

### Testing Device Firmware Upgrade

> Visit
> [FOTA upgrades](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf52.html#fota-upgrades)
> and read `Download the new image to a device` section to see how to upgrade your
> device firmware over Bluetooth LE using smartphone.

Reviewer: ```suggestion
Read the section about downloading the new image to a device on the
[FOTA upgrades](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf52.html#fota-upgrades)
page in the nRF Connect documentation to see how to upgrade your
device firmware over Bluetooth LE using a smartphone.
```
